File: csv_calculator.py
import csv
from typing import Dict, List


def load_sales_data(filename: str) -> List[Dict[str, str]]:
    with open(filename, mode="r") as file:
        reader = csv.DictReader(file)

        return [row for row in reader]


def calculate_total_sales(sales_data: List[Dict[str, str]]) -> Dict[str, float]:
    total_sales = {}

    for row in sales_data:

        product = row["product"]
        quantity = int(row["quantity"])
        price = float(row["price"])

        total_sales[product] = total_sales.get(product, 0) + quantity * price

    return total_sales


def calculate_total_quantity(sales_data: List[Dict[str, str]]) -> Dict[str, int]:
    total_quantity = {}

    for row in sales_data:

        product = row["product"]
        quantity = int(row["quantity"])
        total_quantity[product] = total_quantity.get(product, 0) + quantity

    return total_quantity


def calculate_average_price(sales_data: List[Dict[str, str]]) -> Dict[str, float]:
    total_price = {}
    total_count = {}

    for row in sales_data:

        product = row["product"]
        price = float(row["price"])
        total_price[product] = total_price.get(product, 0) + price
        total_count[product] = total_count.get(product, 0) + 1

    return {
        product: total_price[product] / total_count[product] for product in total_price
    }


def display_results(results: Dict[str, float], title: str) -> None:
    print(f"\n{title}")
    for product, value in results.items():
        print(f"{product}: {value:.2f}")


# Example usage:

filename = "sales.csv"
sales_data = load_sales_data(filename)

total_sales = calculate_total_sales(sales_data)
display_results(total_sales, "Total Sales per Product")

total_quantity = calculate_total_quantity(sales_data)
display_results(total_quantity, "Total Quantity Sold per Product")

average_price = calculate_average_price(sales_data)
display_results(average_price, "Average Price per Product")

File: csv-calculator.js
import { readFileSync } from "fs";

// Example usage:
const data = loadSalesData("sales.csv");

const totalSales = calculateTotalSales(data);
displayResults(totalSales, "Total Sales");
displayResults(calculateTotalQuantity(data), "Total Quantity");
displayResults(calculateAveragePrice(data), "Average Price");
displayResults(calculateHighestTotalSales(totalSales), "Highest Total Sales");
displayResults(calculateAveragePriceWeighted(data), "Weighted Average Price");

function loadSalesData(filename) {
  const data = readFileSync(filename, "utf-8").trim();
  const lines = data.split("\n");

  const headers = lines[0].split(",");
  const salesData = [];

  for (let i = 1; i < lines.length; i++) {
    const values = lines[i].split(",");
    const row = {};

    for (let j = 0; j < headers.length; j++) {
      row[headers[j].trim()] = values[j].trim();
    }

    salesData.push(row);
  }

  return salesData;
}

function calculateTotalSales(salesData) {
  const totalSales = {};

  for (const row of salesData) {
    const product = row.product;
    const quantity = parseInt(row.quantity, 10);
    const price = parseFloat(row.price);

    if (!(product in totalSales)) {
      totalSales[product] = 0;
    }

    totalSales[product] += quantity * price;
  }

  return totalSales;
}

function calculateTotalQuantity(salesData) {
  const totalQuantity = {};

  for (const row of salesData) {
    const product = row.product;
    const quantity = parseInt(row.quantity, 10);

    if (!(product in totalQuantity)) {
      totalQuantity[product] = 0;
    }

    totalQuantity[product] += quantity;
  }

  return totalQuantity;
}

function calculateAveragePrice(salesData) {
  const totalPrice = {};
  const totalCount = {};

  for (const row of salesData) {
    const product = row.product;
    const price = parseFloat(row.price);

    if (!(product in totalPrice)) {
      totalPrice[product] = 0;
      totalCount[product] = 0;
    }

    totalPrice[product] += price;
    totalCount[product] += 1;
  }

  const averagePrice = {};

  for (const product in totalPrice) {
    averagePrice[product] = totalPrice[product] / totalCount[product];
  }

  return averagePrice;
}

function calculateHighestTotalSales(totalSales) {
  const highestTotalSales = {};
  var highestTotalSaleValue = 0;
  var highestTotalSaleProduct;
  
  for (let product in totalSales) {
    if (highestTotalSaleValue < totalSales[product]){
      highestTotalSaleValue = totalSales[product];
      highestTotalSaleProduct = product;
    }
  }

  highestTotalSales[highestTotalSaleProduct] = highestTotalSaleValue;

  return highestTotalSales;
}

function calculateAveragePriceWeighted(salesData) {
  const totalPrice = {};
  const totalCount = {};

  for (const row of salesData) {
    const product = row.product;
    const price = parseFloat(row.price);
    const quantity = parseFloat(row.quantity);

    if (!(product in totalPrice)) {
      totalPrice[product] = 0;
      totalCount[product] = 0;
    }

    totalPrice[product] += (price * quantity);
    totalCount[product] += quantity;
  }

  const averagePrice = {};

  for (const product in totalPrice) {
    averagePrice[product] = totalPrice[product] / totalCount[product];
  }

  return averagePrice;
}

function displayResults(results, title) {
  console.log(`\n${title}`);

  for (const product in results) {
    console.log(`${product}: ${results[product].toFixed(2)}`);
  }
}
